Fix TimeFormat

dlindhol opened the pull request:
This starts with a rewrite of `TimeFormat` with `DateTimeFormatter` instead of `SimpleDataFormat` (which is not thread safe). This should all be safely encapsulated so users of the `TimeFormat` API should not be affected (other than the loss of support for 2-digit century start dates which is not yet being used in latis3).

`DateTimeFormatter` doesn't like to provide precision beyond what the format specifies. For example, you can't get day info from a format using `yyyy-MM`. The new `TimeFormat` will fill the time components with "zeros" to appease the API.

I also made `fromIsoValue` more robust when it comes to interpreting arbitrary time strings as ISO 8601. We used to rely solely on the length of the string and the presence of a "-".

dlindhol committed:
Rewrite TimeFormat in terms of DateTimeFormatter

dlindhol committed:
Interpret time values as ISO first for conversion

JakeMartin99 commented:
I don't know if this is relevant/useful at all, but to your point about filling extra components with 0s I've noticed with the ToHapiTime stuff I've been working on that that code is already doing exactly that (ie any date in any format is successfully extrapolated to a yyyy-MM-dd'T'HH:mm:ss.SSS'Z' format, with higher precision values defaulted to 0 if needed).

dlindhol committed:
Add a Scalar method to get a value as a Double

This enables text time variables to be handled outside of the
core API. It also facilitates any math which uses Double as the
default numeric type.

dlindhol committed:
Support selections with bin semantics for text time variables

dlindhol commented:
I rebased on master to fix the netCDF problems.

dlindhol commented on core/src/main/scala/latis/time/Time.scala:
```diff
@@ -58,37 +59,47 @@ class Time protected (
-    }).leftMap { le =>
-      LatisException(s"Failed to interpret time value: $value", le)
+  override def convertValue(value: String): Either[Exception, Datum] = {
+    TimeFormat.parseIso(value).flatMap { ms =>
```
What about numeric type?
Require quotes?

dlindhol replied:
This is the correct behavior as currently designed. The value will be interpreted as ISO above all other units. The consequence is that users can't reliably select on native numeric units for some values (4, 7, or 8 digits). Note that this could also be the case for some native time *formats* (e.g. "DDDyyyy") that would be interpreted as ISO ("yyyyDDD") over the native units. 

I have a note to review the use of quotes in the future.  We could at least say that unquoted numbers should be treated as the native units if the variable is numeric. We might want to go as far as to require quoting of text variable values, including ISO time values. Not sure what we could do to support ambiguous non-ISO native time formats. Maybe use a `selectNative` operation?
